File: torch/legacy/nn/LogSigmoid.py
import torch
from .Module import Module
from .utils import clear


class LogSigmoid(Module):

    def __init__(self):
        super(LogSigmoid, self).__init__()
        self.buffer = None

    def updateOutput(self, input):
        if self.buffer is None:
            self.buffer = input.new()
        self._backend.LogSigmoid_updateOutput(
            self._backend.library_state,
            input,
            self.output,
            self.buffer
        )
        return self.output

    def updateGradInput(self, input, gradOutput):
        self._backend.LogSigmoid_updateGradInput(
            self._backend.library_state,
            input,
            gradOutput,
            self.gradInput,
            self.buffer
        )
        return self.gradInput

    def clearState(self):
        clear(self, 'buffer')
        return super(LogSigmoid, self).clearState()

File: torch/lib/THNN/generic/LogSigmoid.c
#ifndef TH_GENERIC_FILE
#define TH_GENERIC_FILE "generic/LogSigmoid.c"
#else

void THNN_(LogSigmoid_updateOutput)(
          THNNState *state,
          THTensor *input,
          THTensor *output,
          THTensor *buffer)
{
  THTensor_(resizeAs)(output, input);
  THTensor_(resizeAs)(buffer, input);

  TH_TENSOR_APPLY3(real, output, real, input, real, buffer,
    real z = exp(-*input_data);
    *buffer_data = z;
    *output_data = -log(1. + z);
  );
}

void THNN_(LogSigmoid_updateGradInput)(
          THNNState *state,
          THTensor *input,
          THTensor *gradOutput,
          THTensor *gradInput,
          THTensor *buffer)
{
  THNN_CHECK_NELEMENT(input, gradOutput);
  THTensor_(resizeAs)(gradInput, buffer);
  TH_TENSOR_APPLY3(real, gradInput, real, gradOutput, real, buffer,
    real z = *buffer_data;
    *gradInput_data = *gradOutput_data * z / (1. + z);
  );
}

#endif
